src/main/scala/com/scality/clueso/MetadataIngestionPipeline.scala

```diff
@@ -0,0 +1,187 @@
+    } else {
+      recordNo + compactionRecordInterval - recordNo % compactionRecordInterval
+    }
+  }
+
+  import org.apache.spark.sql.functions.udf
+  val msg_rewrite = udf(msgRewriteFun)
+  val max_op_index = udf(maxOpIndexFun)
+
+  /**
+    * Applies projections and conditions to incoming data frame
+    * Ignores malformed json and garbage payloads
+    * Ignores events from specified bucket name
+    *
+    * @param bucketNameToFilterOut
+    * @param eventStream
+    * @return
+    */
+  def filterAndParseEvents(bucketNameToFilterOut : String, eventStream: DataFrame)(implicit spark : SparkSession, config: CluesoConfig) = {
+
+    var df = eventStream.select(
+      col("timestamp").cast(TimestampType).as("kafkaTimestamp"),
+      msg_rewrite(col("value")).as("content")
+    )
+
+    // defensive filtering to not process kafka garbage
+    df = df.filter(col("content").isNotNull)
+          .filter(length(col("content")).gt(3))
+          .select(
+            col("kafkaTimestamp"),
+            from_json(col("content"), CluesoConstants.eventSchema).alias("event")
+          )
+
+    df = df.filter(col("event").isNotNull.and(col("event.type").isNotNull))
+      .withColumn("key", when(
+        col("event").isNotNull.and(
+          col("event.key").isNotNull
+        ), col("event.key")).otherwise("")
+      )
+      .withColumn("bucket",
+        when(
+          col("event").isNotNull.and(
+            col("event.bucket").isNotNull
+          ), col("event.bucket")).otherwise("NOBUCKET")
+      )
+      .withColumn("type", col("event.type"))
+      .withColumn("opIndex", col("event.opIndex"))
+      .withColumn("maxOpIndex", max_op_index(lit(config.compactionRecordInterval), col("event.opIndex")))
+      .withColumn("message", col("event.value"))
+
+    df = df.filter(
+      !col("bucket").eqNullSafe(bucketNameToFilterOut) &&
+        !col("bucket").eqNullSafe("users..bucket") &&
+        !col("bucket").eqNullSafe("__metastore") &&
+        (col("bucket").isNotNull && !col("bucket").startsWith("mpuShadowBucket")))
+
+      df.drop("event")
+  }
+
+  def main(args: Array[String]): Unit = {
+    require(args.length > 0, "specify configuration file")
+
+    implicit val config = SparkUtils.loadCluesoConfig(args.head)
+
+    // create dir no matter what
+    val fs = SparkUtils.buildHadoopFs(config)
+
+    logger.info(s"Creating directory ${PathUtils.landingURI}")
+    logger.info(s"Creating directory ${PathUtils.stagingURI}")
+
+    fs.mkdirs(new Path(PathUtils.landingURI))
+    fs.mkdirs(new Path(PathUtils.stagingURI))
+
+
+    implicit val spark = SparkUtils.buildSparkSession(config)
+      .appName("Metadata Ingestion Pipeline")
+      .getOrCreate()
+
+    val eventStream = spark.readStream
+      .format("kafka")
+      .option("kafka.bootstrap.servers", config.kafkaBootstrapServers)
+      .option("subscribe", config.kafkaTopic)
+      .option("startingOffsets", "earliest")
+      .option("failOnDataLoss", "false")
```

Comment:
what does this do?

Reply:
From Spark doc: "Whether to fail the query when it's possible that data is lost (e.g., topics are deleted, or offsets are out of range). This may be a false alarm. You can disable it when it doesn't work as you expected."

Basically when we restart the pipeline, the checkpoint could be pointing at offset = 1000 and pipeline attempted to consume from that point onwards... however, Kafka had lost all its data because the volumes in kafka docker container weren't persistent.. and we saw the error.

Disabling this made it work.. but as we are discussing what should go on a docker volume and what doesn't, we can change this accordingly. If this can be set to true, I think we should include this as a application.conf variable because if something goes wrong into production, you may want to be able to start without failOnDataLoss=true.

Reply:
I agree. We should have this configurable. Since we start from earliest offset is there a huge risk if there is "data loss" in this sense?